### ISSUE TITLE
internal server error when viewing scoreboard

### ISSUE THREAD

**Issue by donnm** [UCL-INGI/INGInious#294]
Version v0.4 from git, Debian 9 + lighttpd/fastcgi. Set up a scoreboard, excerpt from `course.yaml`:

```
scoreboard:
    - content: ["primes"]
      reverse: True
      name: "Primes scoreboard (execution time)"
```

Save score in the `run` script using `feedback-custom score $(( $end - $start ))` where end and start are timestamps in milliseconds.

On viewing the scoreboard, I get `internal server error` in the browser and the following in `/var/log/lighttpd/error.log`:

```
(mod_fastcgi.c.2543) FastCGI-stderr: Traceback (most recent call last):                                                                                                                                
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 257, in process                                                                          
(mod_fastcgi.c.2543) FastCGI-stderr:     return self.handle()                                                                                                                                          
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 248, in handle                                                                           
(mod_fastcgi.c.2543) FastCGI-stderr:     return self._delegate(fn, self.fvars, args)                                                                                                                   
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 473, in _delegate                                                                        
(mod_fastcgi.c.2543) FastCGI-stderr:     return handle_class(f)                                                                                                                                        
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 466, in handle_class                                                                     
(mod_fastcgi.c.2543) FastCGI-stderr:     return tocall(*args)                                                                                                                                          
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/pages/utils.py", line 126, in GET                                                        
(mod_fastcgi.c.2543) FastCGI-stderr:     return self.GET_AUTH(*args, **kwargs)                                                                                                                         
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/plugins/scoreboard/__init__.py", line 114, in GET_AUTH                                   
(mod_fastcgi.c.2543) FastCGI-stderr:     for username, userinfo in self.user_manager.get_users_info(users).items():                                                                                    
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/user_manager.py", line 205, in get_users_info                                            
(mod_fastcgi.c.2543) FastCGI-stderr:     infos = method.get_users_info(remaining_users)                                                                                                                
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/plugins/auth/db_auth/__init__.py", line 66, in get_users_info
(mod_fastcgi.c.2543) FastCGI-stderr:     for user in data:
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 1176, in next
(mod_fastcgi.c.2543) FastCGI-stderr:     if len(self.__data) or self._refresh():
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 1087, in _refresh
(mod_fastcgi.c.2543) FastCGI-stderr:     self.__send_message(q)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 924, in __send_message
(mod_fastcgi.c.2543) FastCGI-stderr:     **kwargs)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/mongo_client.py", line 1026, in _send_message_with_response
(mod_fastcgi.c.2543) FastCGI-stderr:     exhaust)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/mongo_client.py", line 1037, in _reset_on_error
(mod_fastcgi.c.2543) FastCGI-stderr:     return func(*args, **kwargs)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/server.py", line 94, in send_message_with_response
(mod_fastcgi.c.2543) FastCGI-stderr:     set_slave_okay, sock_info, use_find_cmd)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/message.py", line 323, in get_message
(mod_fastcgi.c.2543) FastCGI-stderr:     spec, self.fields, self.codec_options)
(mod_fastcgi.c.2543) FastCGI-stderr: bson.errors.InvalidDocument: Cannot encode object: {'<my_ldap_username>'}
```

`<my_ldap_username>` is a placeholder for my username.

**Comment by donnm** [comment]
Logged in as a local user, this gives me the same error. So it's not related to the logged in user, but maybe to the user who is on the scoreboard?

**Comment by donnm** [comment]
Removing all submissions from the task related to this scoreboard gives this trace (same `internal server error` in the browser):
```
(mod_fastcgi.c.2543) FastCGI-stderr: Traceback (most recent call last):
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 257, in process
(mod_fastcgi.c.2543) FastCGI-stderr:     return self.handle()
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 248, in handle
(mod_fastcgi.c.2543) FastCGI-stderr:     return self._delegate(fn, self.fvars, args)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 473, in _delegate
(mod_fastcgi.c.2543) FastCGI-stderr:     return handle_class(f)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 466, in handle_class
(mod_fastcgi.c.2543) FastCGI-stderr:     return tocall(*args)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/pages/utils.py", line 126, in GET
(mod_fastcgi.c.2543) FastCGI-stderr:     return self.GET_AUTH(*args, **kwargs)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/plugins/scoreboard/__init__.py", line 114, in GET_AUTH
(mod_fastcgi.c.2543) FastCGI-stderr:     for username, userinfo in self.user_manager.get_users_info(users).items():
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/user_manager.py", line 205, in get_users_info
(mod_fastcgi.c.2543) FastCGI-stderr:     infos = method.get_users_info(remaining_users)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/inginious/frontend/webapp/plugins/auth/db_auth/__init__.py", line 66, in get_users_info
(mod_fastcgi.c.2543) FastCGI-stderr:     for user in data:
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 1176, in next
(mod_fastcgi.c.2543) FastCGI-stderr:     if len(self.__data) or self._refresh():
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 1087, in _refresh
(mod_fastcgi.c.2543) FastCGI-stderr:     self.__send_message(q)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/cursor.py", line 924, in __send_message
(mod_fastcgi.c.2543) FastCGI-stderr:     **kwargs)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/mongo_client.py", line 1026, in _send_message_with_response
(mod_fastcgi.c.2543) FastCGI-stderr:     exhaust)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/mongo_client.py", line 1037, in _reset_on_error
(mod_fastcgi.c.2543) FastCGI-stderr:     return func(*args, **kwargs)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/server.py", line 94, in send_message_with_response
(mod_fastcgi.c.2543) FastCGI-stderr:     set_slave_okay, sock_info, use_find_cmd)
(mod_fastcgi.c.2543) FastCGI-stderr:   File "/usr/local/lib/python3.5/dist-packages/pymongo/message.py", line 323, in get_message
(mod_fastcgi.c.2543) FastCGI-stderr:     spec, self.fields, self.codec_options)
(mod_fastcgi.c.2543) FastCGI-stderr: bson.errors.InvalidDocument: Cannot encode object: set()
```

**Comment by anthonygego** [comment]
I backported the fix in v0.4 branch as 9d89244b3523eff8c2483fde449b2400aae0b451. Actually a set was passed instead of a list and pymongo cannot serialize it. This only happened using ``db_auth`` plugin. Hopefully by updating the code this should work.

**Comment by GuillaumeDerval** [comment]
Scoreboard only manages, for now, floating points numbers and integers. Feel free to submit a PR to improve this :-)

I'll close this issue as the initial problem is now solved.
